services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java

```diff
@@ -317,6 +321,39 @@ public OIDCEndpoint(AuthenticationCallback callback, RealmModel realm, EventBuil
             super(callback, realm, event);
         }
 
+        @Override
+        public SimpleHttp generateTokenRequest(String authorizationCode) {
```

Comment:
I think makes more sense the PKCE code here `org.keycloak.broker.oidc.AbstractOAuth2IdentityProvider#authenticateTokenRequest`. As this may also be useful for OAuth2-based brokers.

Reply:
I gave this idea a spin, but in that case I'm missing the EventBuilder dependencies. Also IMHO the main purpose of PKCE is to protect the authorizationCode, which is handled in generateTokenRequest(...).

The main purpose of the `authenticateTokenRequest(..)` seems to be to add the proper authentication for the authorization  as well as token refresh requests. Therefore I think the proper place would be `org.keycloak.broker.oidc.AbstractOAuth2IdentityProvider.Endpoint#generateTokenRequest(..)`.

What do you think? I just changed the code to reflect my thinking.